lib/internal/util.js

```diff
@@ -1,18 +1,65 @@
 'use strict';
 
-exports.printDeprecationMessage = function(msg, warned) {
+const util = require('util');
+const warnedLocations = new Set();
+
+function getActualCallerLocation() {
+  // We split three times, because stack trace will be like this
+  //   1. Error
+  //   2. `deprecated` function in `util.deprecate`
+  //   3. The file which actually called `util.deprecate`
+  // and we take the last element after splitting three times
+  const errorObject = {};
+  Error.captureStackTrace(errorObject, exports.printDeprecationMessage);
+  const location = errorObject.stack.split('\n', 3).pop();
+
+  if (/\(.*:\d+:\d+\)/.test(location)) {
+    // If the location is of the following format
+    //     at printer (/home/thefourtheye/Desktop/Test.js:2:15)
+    // we pick only the string within the parens and only till the first :
+    let match = location.match(/\((.*?:\d+):\d+\)/);
+
+    // if match, pick the captured group, otherwise return an empty string
+    return match ? match[1] : '';
+  } else {
+    // If the location is of the following format
+    //     at /home/thefourtheye/Desktop/Test.js:16:5
+    // we remove `at` and take the string excluding the part after last :
+    return location.replace(/\s*at\s*/, '').split(':').slice(0, -1).join(':');
+  }
+
+}
+
+exports.printDeprecationMessage = function(msg, warned, printCaller) {
+  // NOTE: If this function is called directly with a truthy value for
+  // `printCaller` then the filename in the message may not be exactly correct.
+  // `printCaller` will be set to `true` explicitly by `util.deprecate`.
+
   if (process.noDeprecation)
     return true;
 
   if (warned)
     return warned;
 
+  const location = printCaller ? ` (${getActualCallerLocation()})` : '';
+
   if (process.throwDeprecation)
-    throw new Error(msg);
-  else if (process.traceDeprecation)
-    console.trace(msg);
-  else
-    console.error(msg);
+    throw new Error(`(node) ${msg}`);
+  else if (process.traceDeprecation) {
+
+    if (location === '' ||
+        warnedLocations.size !== warnedLocations.add(location).size) {
+      console.trace(`(node) ${msg}`);
+    }
+
+    // if location is empty, then print only once, as we cannot reliably
+    // determine the caller location. If `true` is returned, then next time
+    // `warned` will be `true` and trace will not be printed.
+    return location === '';
+  }
+  else {
+    console.error(`(node) ${msg}${location}`);
```

Comment:
Ah. How did I miss that? I ll change this. I should better add a test with external users case.